client/shared/src/main/scala/org/http4s/client/Client.scala

```diff
@@ -289,16 +291,59 @@ object Client {
       go(source).stream
     }
 
-    def runOrDispose(req: Request[F]): F[Resource[F, Response[F]]] =
-      Ref[F].of(false).map { disposed =>
-        val req0 = addHostHeaderIfUriIsAbsolute(req.pipeBodyThrough(until(disposed)))
-        Resource
-          .eval(app(req0))
-          .onFinalize(disposed.set(true))
-          .map(_.pipeBodyThrough(until(disposed)))
-      }
+    def run(
+        message: Message[F],
+        refOp: Option[Ref[F, Boolean]] = None,
+    ): Resource[F, Response[F]] = message match {
+      case req: Request[F] =>
+        Resource.eval(Ref[F].of(false)).flatMap { disposed =>
+          val reqAugmented =
+            addHostHeaderIfUriIsAbsolute(req.pipeBodyThrough(until(disposed)))
+          Resource
+            .eval(app(reqAugmented))
+            .onFinalize(disposed.set(true))
+            .flatMap(run(_, Option(disposed)))
+        }
+      case resp: Response[F] =>
+        for {
+          disposed <- Resource.eval(refOp.fold(Ref[F].of(false))(F.pure))
+          channel <- Resource.eval(Channel.synchronous[F, Chunk[Byte]])
+
+          _ <- resp.body.chunks
+            .through(channel.sendAll)
+            .compile
+            .drain
+            .background
+
+          ifDisposed = Stream
+            .eval(disposed.get)
+            .ifM(
+              Stream.eval(F.raiseError(new IOException("response was disposed"))),
+              Stream.empty,
+            )
+
+          r = resp.withBodyStream(
+            channel.stream.unchunks
+              .ifEmpty(ifDisposed)
```

Comment:
Should we just check `disposed` first, does it matter if it is empty?

Reply:
We can do it, but I'm not sure if we really need to because the logic drains the body in `onFinalize`, and `channel.stream` won't return anything except an empty stream. Additionally, I believe we can remove this exception in main since using `Channel` makes it impossible to recompile the response, and this exception is only needed to maintain the same contract.

Reply:
Right, I think it's about communicating intent i.e. if `disposed` is `true`, we should raise an error immediately, without doing anything else.

Reply:
I forgot to mention that since `channel.stream` always returns an empty stream, any map or tap operation isn't invoked. Therefore, I found the `.isEmpty` method useful to maintain the contract. In reality, it doesn't do anything else. For the same reason, I removed `until(disposed))` from the response body stream. There is no case where it fires due to the "Channel" logic.

Reply:
I'm wondering if you mean that the whole body stream should be started with `Stream.eval(disposed.get)` check? Like:
```scala
Stream
   .eval(disposed.get)
   .ifM(
       Stream.eval(F.raiseError(new IOException("response was disposed"))),
       channel.stream....,
    )
```

Reply:
Yes, exactly that! Sorry I wasn't clear.